File: src/main/java/net/crtrpt/whisper/Function.java
package net.crtrpt.whisper;

import java.util.List;
import java.util.Map;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;


public class Function {

    private Scope parentScope;
    private List<TerminalNode> params;
    private ParseTree block;


    public Function(Scope parentScope, List<TerminalNode> params, ParseTree block) {
        this.parentScope = parentScope;
        this.params = params;
        this.block = block;
    }

    public TLValue invoke(List<TLValue> args, Map<String, Function> functions, Map<String, BuildInFunction> buildFunction) {
        if (args.size() != this.params.size()) {
            throw new RuntimeException("Illegal Function call");
        }
        Scope scopeNext = new Scope(parentScope, true); // create function scope

        for (int i = 0; i < this.params.size(); i++) {
            TLValue value = args.get(i);
            //获取每个函数的参数值
            scopeNext.assignParam(this.params.get(i).getText(), value);
        }
        EvalVisitor evalVistorNext = new EvalVisitor(scopeNext, functions,buildFunction);

        TLValue ret = TLValue.VOID;
        try {
            evalVistorNext.visit(this.block);
        } catch (ReturnValue returnValue) {
            ret = returnValue.value;
        }
        return ret;
    }
}

File: dart/bin/Function.dart
import 'dart:math';

import 'package:antlr4/antlr4.dart';

import 'EvalException.dart';
import 'EvalVisitor.dart';
import 'ReturnValue.dart';
import 'Scope.dart';
import 'TLValue.dart';

class Function {
  Scope parentScope;
  List<TerminalNode> params;
  ParseTree block;

  Function(this.parentScope, this.params, this.block);

  TLValue invoke(List<TLValue> args, Map<dynamic, dynamic> functions, buildFunction) {
    if (args.length != params.length) {
      throw EvalException(msg: '函数调用出错');
    }

    var scopeNext = Scope(parentScope, true); // create function scope

    for (var i = 0; i < params.length; i++) {
      var value = args[i];
      //获取每个函数的参数值
      scopeNext.assignParam(params[i].text, value);
    }
    var evalVistorNext = EvalVisitor(scopeNext, functions, buildFunction);

    var ret = TLValue.VOID;
    try {
      evalVistorNext.visit(block);
    } on ReturnValue catch (e) {
      ret = e.value;
    }
    return ret;
  }
}
